frontend/src/providers/QueryBuilder.tsx

```diff
@@ -64,37 +61,104 @@ export function QueryBuilderProvider({
 		setQueryBuilderData(initialQueryBuilderData);
 	}, []);
 
-	// ** Method for setupping query builder data
+	// ** Method for setuping query builder data
+	// ** Before setuping transform data from backend to frontend format
 	const initQueryBuilderData = useCallback(
 		(queryBuilderData: QueryBuilderData): void => {
 			setQueryBuilderData(queryBuilderData);
 		},
 		[],
 	);
 
+	const removeEntityByIndex = useCallback(
+		(type: keyof QueryBuilderData, index: number) => {
+			setQueryBuilderData((prevState) => {
+				const currentArray: (IBuilderQueryForm | IBuilderFormula)[] =
+					prevState[type];
+				return {
+					...prevState,
+					[type]: currentArray.filter((item, i) => index !== i),
+				};
+			});
+		},
+		[],
+	);
+
+	const addNewQuery = useCallback(() => {
+		setQueryBuilderData((prevState) => {
+			if (prevState.queryData.length >= MAX_QUERIES) return prevState;
```

Comment:
same here also

Reply:
This is setter from useState. Do you suggest separate logic inside where prevState to another function?

Reply:
this comment is to not to move logic into another in better it would be 


break long function into smaller chunks to increase the readability and testing scope of the function